### PR TITLE
MLIBZ-1908: x-kinvey-custom-request-properties header

### PR DESCRIPTION
**Pull request by heyzooi**
#### Description

Adding `X-Kinvey-Custom-Request-Properties` header if any `client.customRequestProperties` was set

#### Changes

* The property `client.customRequestProperties` already exists but the HTTP request was ignoring it, but now it's considering

#### Tests

* New unit test to check if the header was added to the request

### CODE REVIEW

File: Kinvey/Kinvey/Client.swift [hunk 1]
```diff
@@ -96,7 +96,7 @@ open class Client: Credential {
     open var clientAppVersion: String?
     
     /// Custom request properties for this client instance.
-    open var customRequestProperties: [String : String] = [:]
+    open var customRequestProperties: [String : Any]?
```
Comment (tejasranade) [inline Kinvey/Kinvey/Client.swift]: @heyzooi if the property is called `requestConfiguration` in other places, it should be called something similar here as well. Also worth discussing - if individual methods have the requestConfig option, is there any reason to expose a global config? Overriding, precedence etc adds a lot of (potentially unnecessary) complexity.

File: Kinvey/Kinvey/CustomEndpoint.swift [hunk 2]
```diff
@@ -103,11 +134,18 @@ open class CustomEndpoint {
...
+        params: Params? = nil,
+        client: Client = sharedClient,
+        requestConfiguration: RequestConfiguration? = nil,
+        completionHandler: CompletionHandler<JsonDictionary>? = nil
+    ) -> Request {
         return execute(
             name,
             params: params,
-            client: client
+            client: client,
```
Comment (tejasranade) [inline Kinvey/Kinvey/CustomEndpoint.swift]: @heyzooi just an idea - a majority of methods in the public API have the `client` parameter. What if we make the `client` a part of requestConfig (defaulting to sharedClient)?

File: Kinvey/Kinvey/DataStore.swift [hunk 3]
```diff
@@ -186,8 +186,17 @@ open class DataStore<T: Persistable> where T: NSObject {
      - returns: A `Request` instance which will allow cancel the request later
      */
     @discardableResult
-    open func find(_ id: String, readPolicy: ReadPolicy? = nil, completionHandler: @escaping ObjectCompletionHandler) -> Request {
-        return find(id, readPolicy: readPolicy) { (result: Result<T, Swift.Error>) in
+    open func find(
+        _ id: String,
+        readPolicy: ReadPolicy? = nil,
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: We should deprecate the `readPolicy` parameter and make it part of the `requestConfig`.

File: Kinvey/Kinvey/DataStore.swift [hunk 4]
```diff
@@ -229,8 +249,19 @@ open class DataStore<T: Persistable> where T: NSObject {
      - returns: A `Request` instance which will allow cancel the request later
      */
     @discardableResult
-    open func find(_ query: Query = Query(), deltaSet: Bool? = nil, readPolicy: ReadPolicy? = nil, completionHandler: @escaping ArrayCompletionHandler) -> Request {
-        return find(query, deltaSet: deltaSet, readPolicy: readPolicy) { (result: Result<[T], Swift.Error>) in
+    open func find(
+        _ query: Query = Query(),
+        deltaSet: Bool? = nil,
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: Similar to `readPolicy`, the `deltaSet` parameter can also be in the `requestConfig`.

File: Kinvey/Kinvey/FileStore.swift [hunk 5]
```diff
@@ -98,12 +98,20 @@ open class FileStore<FileType: File> {
     
     /// Uploads a `UIImage` in a PNG or JPEG format.
     @discardableResult
-    open func upload(_ file: FileType, image: NSImage, imageRepresentation: ImageRepresentation = .png, ttl: TTL? = nil, completionHandler: FileCompletionHandler? = nil) -> Request {
+    open func upload(
+        _ file: FileType,
+        image: NSImage,
+        imageRepresentation: ImageRepresentation = .png,
+        ttl: TTL? = nil,
```
Comment (tejasranade) [inline Kinvey/Kinvey/FileStore.swift]: TTL can be part of `requestConfig`.

File: Kinvey/Kinvey/RequestFactory.swift [hunk 6]
```diff
@@ -63,3 +63,22 @@ protocol RequestFactory {
     func buildLiveStreamUnsubscribe(streamName: String, userId: String, deviceId: String) -> HttpRequest
     
 }
+
+//Allow to set a per request
+public struct RequestConfiguration {
```
Comment (tejasranade) [inline Kinvey/Kinvey/RequestFactory.swift]: @heyzooi this makes sense for the `customRequestHeaders` problem you're solving. As my other comments suggest, you should consider using  this requestConfig definition to encapsulate other custom parameters that are exposed on each API.

File: Kinvey/Kinvey/Client.swift (hunk 1 above)
Comment (heyzooi) [inline Kinvey/Kinvey/Client.swift]: good question... I'm going to check with the requester of this feature if make sense to keep a global version

File: Kinvey/Kinvey/CustomEndpoint.swift (hunk 2 above)
Comment (heyzooi) [inline Kinvey/Kinvey/CustomEndpoint.swift]: I think it's a good idea, but it will result in a huge breaking change. Do you think it's worth doing it? I think it's worth it because I'm don't seeing many people using a custom reference for the `client` anyway.

File: Kinvey/Kinvey/DataStore.swift (hunk 3 above)
Comment (heyzooi) [inline Kinvey/Kinvey/DataStore.swift]: I agree, the only catch is `readPolicy` and `writePolicy` are not available for all the requests that requestConfig would be present. We could either subclass `RequestConfiguration` or leave explicit in the docs which options are considered for each call. My opinion: subclass would be better since it make its clear what are the options available.

File: Kinvey/Kinvey/DataStore.swift (hunk 4 above)
Comment (heyzooi) [inline Kinvey/Kinvey/DataStore.swift]: agree, we don't need to decide which approach to do according to the previous comment.

File: Kinvey/Kinvey/FileStore.swift (hunk 5 above)
Comment (heyzooi) [inline Kinvey/Kinvey/FileStore.swift]: totally agree, all those changes will also help to solve this problem here: https://hub.codebeat.co/docs/software-quality-metrics#arity

File: Kinvey/Kinvey/RequestFactory.swift (hunk 6 above)
Comment (heyzooi) [inline Kinvey/Kinvey/RequestFactory.swift]: agree

File: Kinvey/Kinvey/DataStore.swift [hunk 7]
```diff
@@ -186,8 +186,15 @@ open class DataStore<T: Persistable> where T: NSObject {
      - returns: A `Request` instance which will allow cancel the request later
      */
     @discardableResult
-    open func find(_ id: String, readPolicy: ReadPolicy? = nil, completionHandler: @escaping ObjectCompletionHandler) -> Request {
-        return find(id, readPolicy: readPolicy) { (result: Result<T, Swift.Error>) in
+    open func find(
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: @heyzooi  This should be marked deprecated.

File: Kinvey/Kinvey/DataStore.swift [hunk 8]
```diff
@@ -207,11 +214,44 @@ open class DataStore<T: Persistable> where T: NSObject {
      - returns: A `Request` instance which will allow cancel the request later
      */
     @discardableResult
-    open func find(_ id: String, readPolicy: ReadPolicy? = nil, completionHandler: @escaping (Result<T, Swift.Error>) -> Void) -> Request {
+    open func find(
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: This should also be deprecated.

File: Kinvey/Kinvey/DataStore.swift [hunk 9]
```diff
@@ -229,8 +269,17 @@ open class DataStore<T: Persistable> where T: NSObject {
      - returns: A `Request` instance which will allow cancel the request later
      */
     @discardableResult
-    open func find(_ query: Query = Query(), deltaSet: Bool? = nil, readPolicy: ReadPolicy? = nil, completionHandler: @escaping ArrayCompletionHandler) -> Request {
-        return find(query, deltaSet: deltaSet, readPolicy: readPolicy) { (result: Result<[T], Swift.Error>) in
+    open func find(
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: Also needs deprecation.

File: Kinvey/Kinvey/RequestFactory.swift [hunk 10]
```diff
@@ -10,56 +10,97 @@ import Foundation
...
+    func buildLiveStreamSubscribe(streamName: String, userId: String, deviceId: String, options: Options?) -> HttpRequest
+    func buildLiveStreamUnsubscribe(streamName: String, userId: String, deviceId: String, options: Options?) -> HttpRequest
+    
+}
+
+//Allow to set a per request
+public struct Options {
+    
+    public var client: Client?
+    public var clientId: String?
```
Comment (tejasranade) [inline Kinvey/Kinvey/RequestFactory.swift]: @heyzooi why is this necessary?

File: Kinvey/Kinvey/RequestFactory.swift [hunk 11]
```diff
@@ -10,56 +10,97 @@ import Foundation
...
+    func buildOAuthGrantRefreshToken(refreshToken: String, options: Options?) -> HttpRequest
+    
+    func buildLiveStreamGrantAccess(streamName: String, userId: String, acl: LiveStreamAcl, options: Options?) -> HttpRequest
+    func buildLiveStreamPublish(streamName: String, userId: String, options: Options?) -> HttpRequest
+    func buildLiveStreamSubscribe(streamName: String, userId: String, deviceId: String, options: Options?) -> HttpRequest
+    func buildLiveStreamUnsubscribe(streamName: String, userId: String, deviceId: String, options: Options?) -> HttpRequest
+    
+}
+
+//Allow to set a per request
```
Comment (tejasranade) [inline Kinvey/Kinvey/RequestFactory.swift]: Since options is going to be used for more advanced use cases, I suggest writing more detailed API docs.

File: Kinvey/Kinvey/RequestFactory.swift [hunk 12]
```diff
@@ -10,56 +10,97 @@ import Foundation
...
+    public var writePolicy: WritePolicy?
+    public var timeout: TimeInterval?
+    
+    /// App version for this client instance.
+    public var clientAppVersion: String?
+    
+    /// Custom request properties for this client instance.
+    public var customRequestProperties: [String : Any]?
+    
+    public init(
```
Comment (tejasranade) [inline Kinvey/Kinvey/RequestFactory.swift]: @heyzooi Should there be an `init` with no parameters? Since all properties are public, maybe it makes sense to just `init()` and then set the ones you care about?

File: Kinvey/Kinvey/RequestFactory.swift (hunk 10 above)
Comment (heyzooi) [inline Kinvey/Kinvey/RequestFactory.swift]: this is for MIC remember?
